pycbc/events/single.py

```diff
@@ -1,61 +1,141 @@
+                live_time = fit_file[ifo].attrs['live_time']
+                fit_rates = dist_grp['counts'][:] / live_time
+                fit_coeffs = dist_grp['fit_coeff'][:]
+                fit_thresh = fit_file.attrs['fit_threshold']
         return cls(
            ifo, newsnr_threshold=args.single_newsnr_threshold[ifo],
            reduced_chisq_threshold=args.single_reduced_chisq_threshold[ifo],
-           fixed_ifar=args.single_fixed_ifar[ifo],
            duration_threshold=args.single_duration_threshold[ifo],
+           fixed_ifar=args.single_fixed_ifar,
+           fit_bins=fit_bins,
+           fit_coeffs=fit_coeffs,
+           fit_rates=fit_rates,
+           fit_thresh=fit_thresh
            )
 
-    def check(self, triggers, data_reader):
+    def check(self, trigs, data_reader):
         """ Look for a single detector trigger that passes the thresholds in
         the current data.
         """
-        if len(triggers['snr']) == 0:
-            return None
-
-        i = triggers['snr'].argmax()
-        # This uses the pycbc live convention of chisq always meaning the
-        # reduced chisq.
-        rchisq = triggers['chisq'][i]
-        nsnr = ranking.newsnr(triggers['snr'][i], rchisq)
-        dur = triggers['template_duration'][i]
-
-        if nsnr > self.newsnr_threshold and \
-                rchisq < self.reduced_chisq_threshold and \
-                dur > self.duration_threshold:
-            fake_coinc = {'foreground/%s/%s' % (self.ifo, k): triggers[k][i]
-                          for k in triggers}
+        trigsremain = True
+        while trigsremain:
+            if len(trigs['snr']) == 0:
+                trigsremain = False
+                break
+
+            # Apply cuts to trigs before clustering
+            valid_idx = (trigs['template_duration'] >
+                         self.thresholds['duration']) & \
+                        (trigs['chisq'] < self.thresholds['reduced_chisq'])
+            if not np.any(valid_idx):
+                trigsremain = False
+                break
+            cutdurchi_trigs = {k: trigs[k][valid_idx] for k in trigs}
+            nsnr_all = ranking.newsnr(cutdurchi_trigs['snr'],
+                                      cutdurchi_trigs['chisq'])
+            nsnr_idx = nsnr_all > self.thresholds['newsnr']
+            if not np.any(nsnr_idx):
+                trigsremain = False
+                break
+            cutall_trigs = {k: cutdurchi_trigs[k][nsnr_idx]
+                            for k in trigs}
+
+            # 'cluster' by taking the maximal newsnr value over the trigger set
+            i = nsnr_all[nsnr_idx].argmax()
+
+            # This uses the pycbc live convention of chisq always meaning the
+            # reduced chisq.
+            rchisq = cutall_trigs['chisq'][i]
+            nsnr = ranking.newsnr(cutall_trigs['snr'][i], rchisq)
+            dur = cutall_trigs['template_duration'][i]
+            break
```

Comment:
Is the `while` loop appropriate here? On a quick glance, it looks like you will actually never loop around due to the final `break`.

Reply:
The final break is on purpose. This was the way I thought of which would allow breaks so that if all triggers are filtered out, it would be able to break the loop without performing the next bits of code

The other option I can think of would be many nested if statements, or a for loop with just a single loop which can be continued past

Reply:
Okay I've put this bit into a set of if statements which will prevent it from running if all triggers have been cut. This is better practice than the loop